gax-grpc/src/test/java/com/google/api/gax/grpc/GrpcDirectServerStreamingCallableTest.java

```diff
@@ -106,24 +115,116 @@ public void testBadContext() {
 
     CountDownLatch latch = new CountDownLatch(1);
-    MoneyObserver observer = new MoneyObserver(latch);
+    MoneyObserver observer = new MoneyObserver(true, latch);
 
-    Color request = Color.newBuilder().setRed(0.5f).build();
-    streamingCallable.serverStreamingCall(request, observer);
+    thrown.expect(IllegalArgumentException.class);
+    streamingCallable.call(DEFAULT_REQUEST, observer);
   }
 
   @Test
-  public void testServerStreaming() throws Exception {
+  public void testServerStreamingStart() throws Exception {
     CountDownLatch latch = new CountDownLatch(1);
-    MoneyObserver moneyObserver = new MoneyObserver(latch);
+    MoneyObserver moneyObserver = new MoneyObserver(true, latch);
 
-    Color request = Color.newBuilder().setRed(0.5f).build();
-    streamingCallable.serverStreamingCall(request, moneyObserver);
+    streamingCallable.call(DEFAULT_REQUEST, moneyObserver);
+
+    Truth.assertThat(moneyObserver.controller).isNotNull();
+  }
+
+  @Test
+  public void testServerStreaming() throws Exception {
+    CountDownLatch latch = new CountDownLatch(2);
+    MoneyObserver moneyObserver = new MoneyObserver(true, latch);
+
+    streamingCallable.call(DEFAULT_REQUEST, moneyObserver);
 
     latch.await(20, TimeUnit.SECONDS);
     Truth.assertThat(moneyObserver.error).isNull();
-    Money expected = Money.newBuilder().setCurrencyCode("USD").setUnits(127).build();
-    Truth.assertThat(moneyObserver.response).isEqualTo(expected);
+    Truth.assertThat(moneyObserver.response).isEqualTo(DEFAULT_RESPONSE);
+  }
+
+  @Test
+  public void testManualFlowControl() throws Exception {
+    CountDownLatch latch = new CountDownLatch(2);
+    MoneyObserver moneyObserver = new MoneyObserver(false, latch);
+
+    streamingCallable.call(DEFAULT_REQUEST, moneyObserver);
+
+    latch.await(500, TimeUnit.MILLISECONDS);
+    Truth.assertWithMessage("Received response before requesting it")
+        .that(moneyObserver.response)
+        .isNull();
+
+    moneyObserver.controller.request(1);
+    latch.await(500, TimeUnit.MILLISECONDS);
+
+    Truth.assertThat(moneyObserver.response).isEqualTo(DEFAULT_RESPONSE);
+    Truth.assertThat(moneyObserver.completed).isTrue();
+  }
+
+  @Test
+  public void testCancelClientCall() throws Exception {
+    CountDownLatch latch = new CountDownLatch(1);
+    MoneyObserver moneyObserver = new MoneyObserver(false, latch);
+    CancellationException expectedCause = new CancellationException("cancelled!");
+
+    streamingCallable.call(ASYNC_REQUEST, moneyObserver);
+
+    moneyObserver.controller.cancel(expectedCause);
+    moneyObserver.controller.request(1);
+    latch.await(500, TimeUnit.MILLISECONDS);
+
+    Truth.assertThat(moneyObserver.error).isInstanceOf(StatusRuntimeException.class);
+    Truth.assertThat(((StatusRuntimeException) moneyObserver.error).getStatus().getCode())
+        .isEqualTo(Status.CANCELLED.getCode());
+    Truth.assertThat(moneyObserver.error.getCause()).isSameAs(expectedCause);
```

Comment:
I'm wondering what your thoughts are on this behavior. In the current implementation, every passed to onError is a StatusRuntimeException (and in an upcoming PR an ApiException). However, I wondering if it would make more sense in case of cancellation to pass the actual Cancellation exception directly w/o wrapping it.

Reply:
Well when `ApiException` is thrown, the specific type should be https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/CancelledException.java . I think may be easiest for that to be a consistent type instead of different possible types of cancellation exceptions.

Reply:
This differs from the current behavior of UnaryCallables. When you cancel the future from a UnaryCallable, it will set it's internal exception to a new java.util.concurrent.CancellationException. 

Maybe we can solve both issues with:
- remove `cancel(Throwable)` and for user initiated cancellation via  `cancel()`  pass `java.util.concurrent.Cancellation` to onError.
- for system/client initiated cancellation (when stream is idle for too long, or the user provided ResponseObserver throws an exception), we pass the gax CancellationException

Reply:
I added the change to a separate branch:
https://github.com/igorbernstein2/gax-java/tree/ssc-2b-backpressure

If you think this is a good idea, I'll add it to this PR

Reply:
Ahh ok, we should definitely be consistent with UnaryCallable. Let's use what is on your ssc-2b-backpressure branch.

Reply:
ok, cherry-picked it in